libcudacxx/include/cuda/std/detail/libcxx/include/variant

```diff
@@ -629,1043 +746,1443 @@ template <_Trait _DestructibleTrait, size_t _Index, class... _Types>
+      _LIBCUDACXX_TRAIT(
+        is_convertible, decltype(_Operator{}(_CUDA_VSTD::forward<_T1>(__t1), _CUDA_VSTD::forward<_T2>(__t2))), bool),
+      "the relational operator does not return a type which is "
+      "implicitly convertible to bool");
     return _Operator{}(_CUDA_VSTD::forward<_T1>(__t1), _CUDA_VSTD::forward<_T2>(__t2));
   }
 };
 
+struct __variant_binary_visitor
+{
+  template <class _BinaryOp, class _LeftVariant, class _RightVariant>
+  inline _LIBCUDACXX_HIDDEN _LIBCUDACXX_INLINE_VISIBILITY static constexpr auto
+  __visit(const size_t __index, _BinaryOp&& __op, const _LeftVariant& __lhs, const _RightVariant& __rhs)
+  {
+    return __visit(integral_constant<size_t, variant_size_v<_LeftVariant> - 1>{},
+                   __index,
+                   _CUDA_VSTD::forward<_BinaryOp>(__op),
+                   __lhs,
+                   __rhs);
+  }
+
+private:
+  template <size_t _CurrentIndex, class _BinaryOp, class _LeftVariant, class _RightVariant>
+  inline _LIBCUDACXX_HIDDEN _LIBCUDACXX_INLINE_VISIBILITY static constexpr auto
+  __visit(integral_constant<size_t, _CurrentIndex>,
+          const size_t __index,
+          _BinaryOp&& __op,
+          const _LeftVariant& __lhs,
+          const _RightVariant& __rhs)
+  {
+    if (__index == _CurrentIndex)
+    {
+      return __op(_CUDA_VSTD::get<_CurrentIndex>(__lhs), _CUDA_VSTD::get<_CurrentIndex>(__rhs));
+    }
+    return __visit(
+      integral_constant<size_t, _CurrentIndex - 1>{}, __index, _CUDA_VSTD::forward<_BinaryOp>(__op), __lhs, __rhs);
+  }
+
+  template <class _BinaryOp, class _LeftVariant, class _RightVariant>
+  inline _LIBCUDACXX_HIDDEN _LIBCUDACXX_INLINE_VISIBILITY static constexpr auto
+  __visit(integral_constant<size_t, 0>,
+          const size_t __index,
+          _BinaryOp&& __op,
+          const _LeftVariant& __lhs,
+          const _RightVariant& __rhs)
+  {
+    if (__index == 0)
+    {
+      return __op(_CUDA_VSTD::get<0>(__lhs), _CUDA_VSTD::get<0>(__rhs));
+    }
+    // We already checked that every variant has a value, so we should never reach this line
+    _LIBCUDACXX_UNREACHABLE();
+  }
+};
+
 template <class... _Types>
-inline _LIBCUDACXX_INLINE_VISIBILITY
-constexpr bool operator==(const variant<_Types...>& __lhs,
-                          const variant<_Types...>& __rhs) {
-  using __variant_detail::__visitation::__variant;
-  if (__lhs.index() != __rhs.index()) return false;
-  if (__lhs.valueless_by_exception()) return true;
-  return __variant::__visit_value_at(__lhs.index(), __convert_to_bool<equal_to<>>{}, __lhs, __rhs);
+_LIBCUDACXX_INLINE_VISIBILITY constexpr bool operator==(const variant<_Types...>& __lhs, const variant<_Types...>& __rhs)
+{
+  const auto __index = __lhs.index();
+  if (__index != __rhs.index())
+  {
+    return false;
+  }
+  if (__lhs.valueless_by_exception())
+  {
+    return true;
+  }
+
+  return __variant_binary_visitor::__visit(__index, __convert_to_bool<equal_to<>>{}, __lhs, __rhs);
```

Comment:
Why add the more specialized version of this for two variants instead of reusing `__visit_value`? Is this a compilation time optimization?

Reply:
Yeah, there was quite some effort needed to press it into the `__visit_value` overloads. This is both more efficient compile time wise and also avoids use of the function pointer array